db_sql.go

```diff
@@ -43,31 +43,44 @@ const (
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 type sqlUserStoreDB struct {
-	db *sql.DB
+	db             *sql.DB
+	passwordExpiry time.Duration
+}
+
+func (x *sqlUserStoreDB) SetConfig(passwordExpiry time.Duration) error {
+	x.passwordExpiry = passwordExpiry
+	return nil
 }
 
-func (x *sqlUserStoreDB) Authenticate(identity, password string) error {
+func (x *sqlUserStoreDB) Authenticate(identity, password string, authTypeCheck AuthCheck) (string, error) {
 	row := x.db.QueryRow(`SELECT userid FROM authuserstore WHERE (LOWER(email) = $1 OR LOWER(username) = $1) AND (archived = false OR archived IS NULL)`, CanonicalizeIdentity(identity))
 	var userId int64
 	if err := row.Scan(&userId); err != nil {
-		return ErrIdentityAuthNotFound
+		return "", ErrIdentityAuthNotFound
 	}
-	row = x.db.QueryRow(`SELECT password FROM authuserpwd WHERE userid = $1`, userId)
+
+	row = x.db.QueryRow(`SELECT password, updated FROM authuserpwd WHERE userid = $1`, userId)
 	dbHash := ""
-	if err := row.Scan(&dbHash); err != nil {
-		return ErrIdentityAuthNotFound
+	var lastUpdated time.Time
+	if err := row.Scan(&dbHash, &lastUpdated); err != nil {
+		return "", ErrIdentityAuthNotFound
 	}
 
-	if verifyAuthausHash(password, dbHash) {
-		return nil
-	} else {
-		return ErrInvalidPassword
+	if !verifyAuthausHash(password, dbHash) {
+		return "", ErrInvalidPassword
+	}
+
+	if x.passwordExpiry != 0 && authTypeCheck&AuthCheckPasswordExpired == 2 {
+		if lastUpdated.Add(x.passwordExpiry).Before(time.Now()) {
+			return ErrPasswordExpired.Error(), nil
```

Comment:
This is not done right IMO. You can return this error as the second argument (which is the standard Go way) and then react differently to it upstream.